ext/openssl/ossl_pkey.c

```diff
@@ -446,6 +446,174 @@ pkey_generate(int argc, VALUE *argv, VALUE self, int genparam)
+    if(NIL_P(value))
+        return ST_CONTINUE;
+
+    if (SYMBOL_P(key))
+        key = rb_sym2str(key);
+
+    const char *key_ptr = StringValueCStr(key);
+    const struct pkey_from_parameters_arg *params = (const struct pkey_from_parameters_arg *) arg;
+
+    for(int i = 0; strlen(params->aliases[i].alias) > 0; i++) {
+        if(strcmp(params->aliases[i].alias, key_ptr) == 0) {
+            key_ptr = params->aliases[i].param_name;
+            break;
+        }
+    }
+
+    for (const OSSL_PARAM *settable_params = params->settable_params; settable_params->key != NULL; settable_params++) {
+        if(strcmp(settable_params->key, key_ptr) == 0) {
+            switch (settable_params->data_type) {
+            case OSSL_PARAM_INTEGER:
+            case OSSL_PARAM_UNSIGNED_INTEGER:
+                if(!OSSL_PARAM_BLD_push_BN(params->param_bld, key_ptr, GetBNPtr(value))) {
+                    OSSL_PARAM_BLD_free(params->param_bld);
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_BN");
+                }
+                break;
+            case OSSL_PARAM_UTF8_STRING:
+                StringValue(value);
+                if(!OSSL_PARAM_BLD_push_utf8_string(params->param_bld, key_ptr, RSTRING_PTR(value), RSTRING_LENINT(value))) {
+                    OSSL_PARAM_BLD_free(params->param_bld);
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_utf8_string");
+                }
+                break;
+
+            case OSSL_PARAM_OCTET_STRING:
+                StringValue(value);
+                if(!OSSL_PARAM_BLD_push_octet_string(params->param_bld, key_ptr, RSTRING_PTR(value), RSTRING_LENINT(value))) {
+                    OSSL_PARAM_BLD_free(params->param_bld);
+                    ossl_raise(ePKeyError, "OSSL_PARAM_BLD_push_octet_string");
+                }
+                break;
+            case OSSL_PARAM_UTF8_PTR:
+            case OSSL_PARAM_OCTET_PTR:
+                OSSL_PARAM_BLD_free(params->param_bld);
+                ossl_raise(ePKeyError, "Unsupported parameter \"%s\", handling of OSSL_PARAM_UTF8_PTR and OSSL_PARAM_OCTET_PTR not implemented", key_ptr);
+                break;
+            }
+
+            return ST_CONTINUE;
+        }
+    }
+    OSSL_PARAM_BLD_free(params->param_bld);
+
+    char message_buffer[512] = { 0 };
+    char *cur = message_buffer;
+    char *end = message_buffer + sizeof(message_buffer);
+    for (const OSSL_PARAM *settable_params = params->settable_params; settable_params->key != NULL; settable_params++) {
+        const char *fmt = cur == message_buffer ? "%s" : ", %s";
+        if (cur > end)
+            break;
+        cur += snprintf(cur, end-cur, fmt, settable_params->key);
+    }
+
+    for(int i = 0; strlen(params->aliases[i].alias) > 0; i++) {
+        const char *fmt = cur == message_buffer ? "%s" : ", %s";
+        if (cur > end)
+            break;
+        cur += snprintf(cur, end-cur, fmt, params->aliases[i].alias);
+    }
```

Comment:
The use of snprintf() makes me a bit nervous. Could you use Ruby String for building the error message? Ruby exposes a number of functions for interacting with string, for example:

```c
// Creates an empty string
VALUE str = rb_str_new(NULL, 0);
// Append to `str` using printf-style format
rb_str_catf(str, "%s", params->aliases[i].alias);
```

`ossl_raise()` supports `"%"PRIsVALUE`  format specifier for including Ruby object in the message (as apposed to `"%s"` for a C string).

Reply:
Thanks for the pointer (and the patience), this is my first really touching a bigger gem with C extensions. So learning here.

Reply:
I ended up using an array and joining with a comma as I probably would have done in Ruby also. This to avoid the complexity of checking if a comma needs to be appended or not. Think the result is pretty tidy.